File: src/main/java/com/planbase/pdf/layoutmanager/Padding.java
// Copyright 2013-03-03 PlanBase Inc. & Glen Peterson
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
// http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

package com.planbase.pdf.layoutmanager;

/**
 Represents minimum spacing of the top, right, bottom, and left sides of PDF Page Items.
 */
public class Padding {

    /**
     Default padding of 1.5, 1.5, 2. 1.5 (top, right, bottom, left)
     */
    public static final Padding DEFAULT_TEXT_PADDING = new Padding(1.5, 1.5, 2, 1.5);
    public static final Padding NO_PADDING = new Padding(0, 0, 0, 0);

    private final double top;
    private final double right;
    private final double bottom;
    private final double left;
    private Padding(final double t, final double r, final double b, final double l) {
        top = t; right = r; bottom = b; left = l;
    }
    /** Like HTML it's top, right, bottom, left */
    public static Padding of(double t, double r, double b, double l) {
        if ((t == 0) && (r == 0) && (b == 0) && (l == 0)) { return NO_PADDING; }
        if ((t == 1.5) && (r == 1.5) && (b == 2) && (l == 1.5)) {
            return DEFAULT_TEXT_PADDING;
        }
        return new Padding(t, r, b, l);
    }
    /** Sets all padding values equally */
    public static Padding of(double a) {
        if (a == 0) { return NO_PADDING; }
        return new Padding(a,a,a,a);
    }

    public Dim topLeftPadDim() { return Dim.of(left, top); }
    public Dim botRightPadDim() { return Dim.of(right, bottom); }

    public Dim subtractFrom(Dim outer) {
        return Dim.of(outer.getWidth() - (left + right),
                      outer.getHeight() - (top + bottom));
    }

    public Dim addTo(Dim outer) {
        return Dim.of(outer.getWidth() + (left + right),
                      outer.getHeight() + (top + bottom));
    }

    public Coord applyTopLeft(Coord orig) {
        return Coord.of(orig.getX() + left, orig.getY() - top);
    }

//    public XyOffset topLeftPadOffset() { return XyOffset.of(left, -top); }
//    public XyOffset botRightPadOffset() { return XyOffset.of(right, -bottom); }

    @Override public boolean equals(Object other) {
        // Cheapest operations first...
        if (this == other) { return true; }
        if ( !(other instanceof Padding) ) { return false; }
        // Details...
        final Padding that = (Padding) other;

        return (this.top == that.top) &&
               (this.right == that.right) &&
               (this.bottom == that.bottom) &&
               (this.left == that.left);
    }

    @Override public int hashCode() {
        // There is generally no reason to ever have padding amounts that differ by less than 0.1
        // which in HTML-land would be a tenth of a pixel.  Since we're adding four roughly equal
        // values, if only one is off by 0.1, that becomes 0.025, hence multiplying times 40 instead
        // of times 10.
        return (int) ((top + right + bottom + left) * 40);
    }

    @Override
    public String toString() {
        if ( (top == right) && (top == bottom) && (top == left) ) {
            return "Padding(" + top + ")";
        }
        return "Padding(t=" + top + ", r=" + right + ", b=" + bottom + ", l=" + left + ")";
    }

    public double top() { return top; }
    public double right() { return right; }
    public double bottom() { return bottom; }
    public double left() { return left; }
}

File: src/main/java/com/planbase/pdf/layoutmanager/Padding.kt
// Copyright 2017 PlanBase Inc.
//
// This file is part of PdfLayoutMgr2
//
// PdfLayoutMgr is free software: you can redistribute it and/or modify
// it under the terms of the GNU Affero General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// PdfLayoutMgr is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU Affero General Public License for more details.
//
// You should have received a copy of the GNU Affero General Public License
// along with PdfLayoutMgr.  If not, see <https://www.gnu.org/licenses/agpl-3.0.en.html>.
//
// If you wish to use this code with proprietary software,
// contact PlanBase Inc. <https://planbase.com> to purchase a commercial license.

package com.planbase.pdf.layoutmanager

/**
 * Represents minimum spacing of the top, right, bottom, and left sides of PDF Page Items.
 */
data class Padding(val top: Float,
                   val right: Float,
                   val bottom: Float,
                   val left: Float) {

    /** Sets all padding values equally  */
    constructor(a:Float) : this(a, a, a, a)

    fun topLeftPadDim(): XyDim = XyDim(left, top)

    fun botRightPadDim(): XyDim = XyDim(right, bottom)

    fun subtractFrom(outer: XyDim): XyDim =
            XyDim(outer.width - (left + right),
                  outer.height - (top + bottom))

    fun addTo(outer: XyDim): XyDim =
            XyDim(outer.width + (left + right),
                  outer.height + (top + bottom))

    fun applyTopLeft(orig: XyOffset): XyOffset = XyOffset(orig.x + left, orig.y - top)

    //    public XyOffset topLeftPadOffset() { return XyOffset(left, -top); }
    //    public XyOffset botRightPadOffset() { return XyOffset(right, -bottom); }

    companion object {
        /**
         * Default padding of 1.5, 1.5, 2. 1.5 (top, right, bottom, left)
         */
        val DEFAULT_TEXT_PADDING = Padding(1.5f, 1.5f, 2f, 1.5f)
        val NO_PADDING = Padding(0f, 0f, 0f, 0f)
    }
}
